hw11_cache/src/main/java/ru/otus/cachehw/MyCache.java

```diff
@@ -0,0 +1,62 @@
+package ru.otus.cachehw;
+
+import java.lang.ref.SoftReference;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.WeakHashMap;
+
+public class MyCache<K, V> implements HwCache<K, V> {
+
+    private WeakHashMap<K, V> cache = new WeakHashMap();
+
+    private List<SoftReference<HwListener>> listeners = new ArrayList<>();
+
+    @Override
+    public void put(K key, V value) {
+        cache.put(key, value);
+        notifyAllListeners(key, value, "put");
+    }
+
+    @Override
+    public void remove(K key) {
+        V removedValue = cache.remove(key);
+        notifyAllListeners(key, removedValue, "remove");
+    }
+
+    private void notifyAllListeners(K key, V value, String action) {
+        listeners.forEach(ref -> {
+            if (ref.get()!=null) {ref.get().notify(key,value,"put");}
+        });
```

Comment:
Я бы сохранил listener в переменной

```suggestion
        listeners.forEach(ref -> {
            HwListener listener = ref.get();
            if (listener != null) {
                listener.notify(key, value, "put");
            }
        });
```

Конкретно здесь разницы никакой. Но рекомендую в таких случаях придерживаться этого стиля, т.к. в `ref.get()` могут быть какие-то сложные вычисления для получения объекта.